### ISSUE TITLE
Identical import inside+outside of `TYPE_CHECKING` block does not error

### ISSUE THREAD

**Issue by injust** [astral-sh/ruff#17555]
### Summary

```python
from __future__ import annotations

from collections.abc import AsyncGenerator, Iterable
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator, Iterable

async def foo(bar: Iterable[int]) -> AsyncGenerator[int]:
    pass
```

https://play.ruff.rs/5b99b487-12b4-45f7-b212-b8e5bdfc63bd

The `from collections.abc` import is repeated inside and outside the `if TYPE_CHECKING` block. I expected this to trigger some combination of [TC003](https://docs.astral.sh/ruff/rules/typing-only-standard-library-import/) and/or [F811](https://docs.astral.sh/ruff/rules/redefined-while-unused/), but Ruff doesn't report any errors.

### Version

0.11.6

**Comment by ntBre** [comment]
Thanks for the report! After looking a bit at both rules, I'm not exactly sure which rule the bug is in, but it does look like a bug to me.

**Comment by Daverball** [comment]
This does seem like it should trigger `F811`. Triggering `TC003` doesn't really make sense to me, since the reference points to the typing only import, not the original import.

I assume `F811` doesn't trigger because the second import happens inside a conditional block, which usually means that ruff can't treat the earlier binding as unused, since the conditional binding might not happen. But since `if TYPE_CHECKING` is special, it probably shouldn't be part of this exception.

There are however more questionable permutations of the same problem:

```python
from __future__ import annotations

from collections.abc import Iterable
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterable

def foo(bar: Iterable[int]) -> None:
    assert isinstance(bar, Iterable)
```

Now we have a runtime use of `Iterable`, so `TC004` will trigger, but if we change the semantics for `F811` it will also trigger. So now the two violations/fixes kind of contradict one another (although it's not that bad, since whichever order the fixes happens in, you still end up with the same result: a single import outside the type checking block).

**Comment by Daverball** [comment]
There is however also a world, where adding a new `TC` scoped rule to catch this sort of duplicate import makes more sense, than changing the scope of existing rules. I haven't really thought through the implications of changing how `F811` interacts with redefined imports inside a `if TYPE_CHECKING` block. But I can't think of any real problems off the top of my head. There are however problems for the opposite case, where a typing only import gets shadowed by a runtime import.

**Comment by ntBre** [comment]
I don't have much to add, I think you summarized the issues and options very well!

I think this is technically a duplicate of #12270, which I didn't notice before, and also related to #5952, but if anything I'd probably mark those as the duplicates in light of the additional context here.

**Comment by Daverball** [comment]
I think #5952 has actually been fixed and can be closed, the `default` parameter on `TypeVar` is being visited as a type expression in the current version of Ruff (I assume this has been the case at least since Ruff added support for Python 3.13). But it looks very much like the opposite problem I want to avoid.

**Comment by ntBre** [comment]
Oh good catch, I didn't find the exact commit/PR but it looks like #5952 was fixed between 0.1.11 and 0.1.12.

**Comment by wgmo** [comment]
(To avoid the confusion which I just experienced reading the bottom of this comment thread, ...)

> Oh good catch, I didn't find the exact commit/PR but it looks like this was fixed between 0.1.11 and 0.1.12.

The "this" here isn't _this_ issue #17555, it's #5952 as mentioned [here](https://github.com/astral-sh/ruff/issues/17555#issuecomment-2824698993)

The problem described here, [in this issue](#17555) is still present in ruff 0.11.13!

**Comment by ntBre** [comment]
Thanks, will update my comment!

**Comment by hunterhogan** [comment]
I'm having the same experience. 

# A relatively precise description of the experience

1. Relevant settings
    1. lint.select: [typing-only-standard-library-import (TC003)](https://docs.astral.sh/ruff/rules/typing-only-standard-library-import/#typing-only-standard-library-import-tc003)
    2. lint.select: [redefined-while-unused (F811)](https://docs.astral.sh/ruff/rules/redefined-while-unused/#redefined-while-unused-f811)
2. If a standard library import is only used as a type annotation in a module, but the import isn't defined in a type-checking block, Ruff emits diagnostic TC003 as expected.
3. Ruff will emit diagnostic F811, as expected, if
    1. A standard library import is defined twice,
    2. The scopes of the two definitions are the same, and
    3. The standard library import is not "used" after the first definition and before the second definition.
4. If a standard library import is
    1. defined exactly one time in a type-checking block,
    2. defined exactly one time in the top level of the module (_cf._ [PLC0415](https://docs.astral.sh/ruff/rules/import-outside-top-level/#import-outside-top-level-plc0415)), and
    3. only used as a type annotation in the module, 
    4. then 
        1. Ruff does not emit diagnostic TC003 because the standard library import is defined in a type-checking block, and
        2. Ruff does not emit diagnostic F811 because the two identifier definitions do not have the same scope.

# Summary of the issues and options

Eh, I don't understand all of the technical issues well enough to summarize them.

# A thought

I _feel_ the behavior of TC003 could be a little more aligned with the intent of the rule. Currently, emit the diagnostic if:

1. standard library import,
2. import only used as a type annotation, and
3. import not defined in type-checking block.

I _feel_ that tweaking the third condition would make it a little more aligned. Emit the diagnostic if:

1. standard library import,
2. import only used as a type annotation, and
3. import is defined as a run-time import.

The easiest way to import without importing at run time is a type-checking block.

## What is a run-time import?

I don't know how to categorize a run-time import, but it is possible, right? Pyright's accessibility analysis divides all statements into might-run and never-run categories, right?

### Scope definition heuristic

Y'all already define the scope for all symbols, right? So, you have a scope definition for the symbols defined in the TYPE_CHECKING ast.If.body. Since `TYPE_CHECKING` is an alias for `False`, I imagine that those scope definitions have at least one unique characteristic. (I'm not a programmer, so I'm struggling to explain this clearly.) Pretend every scope definition includes an IP address. Let's say that the scope definition of all symbols defined in a TYPE_CHECKING block always evaluates to an IP address in 10.0.0.0/8. Therefore, define a "run-time import" as anything not in 10.0.0.0/8. It's a heuristic, of course, and not comprehensive.

### Top-level heuristic

Another alternative: y'all already categorize imports as top-level or not top-level. So, emit the diagnostic if:

1. standard library import,
2. import only used as a type annotation, and
3. import is defined as a top-level import.

---

In any case, I hope the first section is helpful. Defining what _a thing is_ can make it easier to define what it _ought to be_. That's a corollary of 千里之行, 始于足下 "The journey of a thousand miles starts from beneath your feet." To travel to a different place, you must first know where you are.

Oh! But this isn't a philosophical issue for me. I have a private package (or I'd share it) that crashes if some of the type annotation symbols are imported at run time. Sometimes, the IDE or an AI agent will add the symbol as a top-level import even though it's already in the TYPE_CHECKING block. Ruff, ty, and Pylance don't flag the situation, so if I don't see it, I end up pushing bad code. 

Despite my odd situation, this should almost certainly be a low priority for y'all. Good luck!

**Comment by ntBre** [comment]
I believe this should be resolved in the next release by the changes in https://github.com/astral-sh/ruff/pull/22560!
